RFC_STORE_MODIFIER.md

```diff
@@ -0,0 +1,185 @@
+- Small surface area
+- Aligned with `react-redux` v6 approach for preparing for concurrent-safeness
+- SSR-support
+
+### Example usage
+
+(Some imports etc omitted for brevity)
+
+```jsx
+import { withModifiedStore } from 'react-redux'
+import reducer from './namesReducer'
+
+function Greeter({ currentName }) {
+  return `Hi ${currentName}!`
+}
+
+const mapStateToProps = state => ({
+  currentName: state.names.currentName
+})
+
+// Structure of this object is not prescriptive,
+// it depends entirely on the implementation of
+// modifyStore
+// In a sense this is quite like an action, in that it
+// is an object that should describe the store modification
+const modification = {
+  newReducers: {
+    names: reducer
+  }
+}
+
+export default withModifiedStore(modification)(
+  connect(mapStateToProps)(Greeter)
+)
+```
+
+```jsx
+import staticReducers from './staticReducers'
+
+// Example simple userland implementation
+let reducers = { ...staticReducers }
+function modifyStore({ newReducers }, storeProxy, state) {
+  reducers = { ...reducers, ...newReducers }
+  storeProxy.replaceReducer(combineReducers(reducers))
+}
+
+const DynamicGreeter = React.lazy(() => import('./Greeter'))
+
+ReactDOM.render(
+  <Provider store={store} modifyStore={modifyStore}>
+    <Suspense fallback={<div>Loading...</div>}>
+      <DynamicGreeter />
+    </Suspense>
+  </Provider>
+)
+```
+
+### API
+
+#### `modifyStore(modification, storeProxy, state)`
+
+This is a callback that is supposed to be implemented by the user/by libraries and passed into `Provider` or directly into `StoreModifier` or `withModifiedStore`.
+
+Most implementations of `modifyStore` need to be stateful, and the only way to support this for SSR is to create one stateful instance per request. Passing it into `<Provider>` puts the function on the context behind the scenes and is the recommended approach even if SSR is not needed.
+
+**Arguments (automatically passed in to the function)**
+
+1. `modification` is an object that can contain any data that needs to be passed from the calling site (`<StoreModifier>`), for example reducers to be added to the store. There is no prescribed shape for this object, it is supposed to differ between implementations
+
+2. `storeProxy` here is a wrapped version of `store` that spies on `replaceReducer` to be able to safely patch the context with the new state if it is called. Maybe this should be a stripped down version of `store`, leaving out `subscribe` and possibly other methods unsafe to use here? Could also keep them in but provide warnings.
+
+3. `state` is passed in, in case it is needed for conditional logic, because `getState` is kind of unsafe here. This `state`-object contains the current state _stored on context for this render_ and is not the latest version of the state in the store.
```

Comment:
Well, it *can be*, but is not necessarily. I can imagine there being cases where you DO want to get the latest store state and not the one on this render. I hesitate changing the semantics of `store.getState()`, I think it would be confusing if it did not always get the latest state from store? (Unless we made it really clear that `storeProxy` is a proxy that works differently, but if we start changing the behaviour of functions to be "context-safe", the expectation would probably be that *all* functions would be safe, something I don't think we can guarantee)

I do agree it would be nice if you could get the resulting state from `replaceReducer` though. `store.getState()` after calling `replaceReducer` returns updated state, but that might contain other changes as well.

I think there is a tradeoff we need to consider carefully here, between "just use store and we'll try to patch things up magically behind the scenes" and being more explicit about what is happening. I'm starting to lean towards providing `replaceReducer` as a separate parameter instead so it is clear it is special, but also provide the unmodified `store` as an extra parameter for those that need it, but make it clear it is potentially unsafe to use. If it is clear to the user that `replaceReducer` is special, it could also return the new "safe" state.

Reply:
IMO, the [store has 4 functions](https://redux.js.org/api/api-reference#store-api) by default: `getState`, `dispatch`, `subscribe` and `replaceReducer`.

- `getState`: not "context-safe"
- `dispatch`: "context-safe" so long as you don't try to use the updated state
- `subscribe`: not "context-safe" but I agree we should strip it out of the `storeProxy` anyway
- `replaceReducer`: "context-safe" if `getState` returns the patched state after the reducer is added

I feel that adding the unmodified `store` but noting it's _unsafeness_ no better than accessing it from the "private" context and it will be easy to abuse and get wrong and become a footgun for users.